plugins.d/alarm-notify.sh

```diff
@@ -433,268 +391,43 @@ filter_recipient_by_criticality() {
+        [ "${x}" = "silent" -o "${x}" = "disabled" ] && continue
+
+        role_recipients="role_recipients_${method_name}"
+        role_recipients=${!role_recipients}
+        default_recipient="DEFAULT_RECIPIENT_$(echo ${method_name} | awk '{print toupper($0)}')"
+
+        a="${role_recipients[${x}]}"
+        [ -z "${a}" ] && a="${!default_recipient}"
+        for r in ${a//,/ } ; do
+            [ "${r}" != "disabled" ] && filter_recipient_by_criticality email "${r}" && arr[${r/|*/}]="1"
+        done
     done
 
-done
-
-# build the list of slack recipients (channels)
-to_slack="${!arr_slack[*]}"
-[ -z "${to_slack}" ] && SEND_SLACK="NO"
-
-# build the list of Microsoft team recipients (channels)
-to_msteam="${!arr_msteam[*]}"
-[ -z "${to_msteam}" ] && SEND_MSTEAM="NO"
-
-# build the list of rocketchat recipients (channels)
-to_rocketchat="${!arr_rocketchat[*]}"
-[ -z "${to_rocketchat}" ] && SEND_ROCKETCHAT="NO"
-
-# build the list of alerta recipients (channels)
-to_alerta="${!arr_alerta[*]}"
-[ -z "${to_alerta}" ] && SEND_ALERTA="NO"
-
-# build the list of flock recipients (channels)
-to_flock="${!arr_flock[*]}"
-[ -z "${to_flock}" ] && SEND_FLOCK="NO"
-
-# build the list of discord recipients (channels)
-to_discord="${!arr_discord[*]}"
-[ -z "${to_discord}" ] && SEND_DISCORD="NO"
-
-# build the list of pushover recipients (user tokens)
-to_pushover="${!arr_pushover[*]}"
-[ -z "${to_pushover}" ] && SEND_PUSHOVER="NO"
-
-# build the list of pushbulet recipients (user tokens)
-to_pushbullet="${!arr_pushbullet[*]}"
-[ -z "${to_pushbullet}" ] && SEND_PUSHBULLET="NO"
-
-# build the list of twilio recipients (phone numbers)
-to_twilio="${!arr_twilio[*]}"
-[ -z "${to_twilio}" ] && SEND_TWILIO="NO"
-
-# build the list of hipchat recipients (rooms)
-to_hipchat="${!arr_hipchat[*]}"
-[ -z "${to_hipchat}" ] && SEND_HIPCHAT="NO"
-
-# build the list of messagebird recipients (phone numbers)
-to_messagebird="${!arr_messagebird[*]}"
-[ -z "${to_messagebird}" ] && SEND_MESSAGEBIRD="NO"
-
-# build the list of kavenegar recipients (phone numbers)
-to_kavenegar="${!arr_kavenegar[*]}"
-[ -z "${to_kavenegar}" ] && SEND_KAVENEGAR="NO"
-
-# check array of telegram recipients (chat ids)
-to_telegram="${!arr_telegram[*]}"
-[ -z "${to_telegram}" ] && SEND_TELEGRAM="NO"
-
-# build the list of pagerduty recipients (service keys)
-to_pd="${!arr_pd[*]}"
-[ -z "${to_pd}" ] && SEND_PD="NO"
-
-# build the list of fleep recipients (conversation webhooks)
-to_fleep="${!arr_fleep[*]}"
-[ -z "${to_fleep}" ] && SEND_FLEEP="NO"
-
-# build the list of custom recipients
-to_custom="${!arr_custom[*]}"
-[ -z "${to_custom}" ] && SEND_CUSTOM="NO"
+    # build the list of recipients
+    # This needs some creativity because email recipients need to be
+    # separated differently from how the others are.
+    oldifs="$IFS"
+    IFS='	'
+    if [ ${method_name} = "email" ] ; then
+        declare to_${method_name}="$(echo ${!arr[*]} | cut -d '	' --output-delimiter=', ' -f 1-)"
```

Comment:
I think you should attempt to use the bash way for this, to avoid forks, pipes, etc.

Reply:
Unless I'm mistaken, there's no way to do this in bash without calling an external command.  `cut` happens to be slightly more efficient in this case than `tr` here.  In theory though, we don't need the call for anything but email (because everything else is just iterating over the list instead of using it directly, and tabs work to separate items for that).

Reply:
What you want to do with this?

in general you can use the following:

convert the associative array keys to a simple array:

```sh
tmp=(  "${!arr[@]}" )
```

Then use `tmp` to get the array members you need:

```sh
start=1
len=${#tmp[@]}
members=$(( len - start ))
methods=( "${tmp[@]:$start:$members}" )
```

So, `${methods[@]}` will have the members you picked.

Reply:
I could also rewrite it to use the same method that's already being used for email addresses.  I was just trying to avoid yet another loop here (it would end up being a conditional loop, but still), while still accounting for spaces being valid in email addresses (tabs technically are too, but nobody uses them at all).  I'm going to do some quick testing to see if it's actually more efficient done with the loop or done with `cut` (the fork may be a deciding factor, but we're forking all over the place anyway in here).

Reply:
Hmm, kind of surprises me a bit, but the loop solution currently used in `alarm-notify.sh` for this seems to be the fastest until you get well above 100 recipient addresses (which is not likely to be a common case, and will probably be rejected by many email servers).  I'll update things to use that then instead of calling out to a subshell and cut.